### PR TITLE
mgr/orch: allow '.' char in hostname

### PR DESCRIPTION
**Pull request by mgfritch**
Fixes: https://tracker.ceph.com/issues/45399
Signed-off-by: Michael Fritch <mfritch@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/orchestrator/_interface.py [hunk 1]
```diff
@@ -1313,19 +1313,22 @@ def _match():
                 # TODO: can a DaemonDescription exist without a hostname?
                 raise err
 
-            if self.hostname == self.daemon_id:
-                # daemon_id == "hostname"
+            # use the bare hostname, not the FQDN.
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:1316]: painful. I don't think that we'll get `DaemonDescription.service_id()` working reliably. 

Looks like we have to make this explicit, as we're doing the OSD - OSDSpec linkage.

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -1313,19 +1313,22 @@ def _match():
...
+            host = self.hostname.split('.')[0]
```
Comment (jmolmo) [inline src/pybind/mgr/orchestrator/_interface.py:1317]: We are changing the hostname... again .... when we should use the bare name and when the FQDN?. 
This is a small snow ball running down hill... (well.. it starts to become bigger)
Take a look to: https://github.com/ceph/ceph/pull/35274

Comment (mgfritch) [reply src/pybind/mgr/orchestrator/_interface.py:1317]: > We are changing the hostname... again .... when we should use the bare name and when the FQDN?.

Yes, but we made a decision to use bare hostname for the the daemon id. Do you have another suggestion how to fix this specific tracker?

> This is a small snow ball running down hill... (well.. it starts to become bigger)
> Take a look to: #35274

maybe we build :snowman: ?!

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:1317]: if we change the name of daemons, we have to cope with both types for at least ceph Q anyway.

**Comment by tchaikov** [comment]
```
2020-06-02T13:20:34.213 INFO:ceph.mon.a.smithi013.stdout:Jun 02 13:20:34 smithi013 bash[7471]: audit 2020-06-02T13:20:33.431015+0000 mon.a (mon.0) 177 : audit [DBG] from='client.? v1:172.21.15.79:0/3533396549' entity='client.admin' cmd=[{
"prefix": "mon dump", "format": "json"}]: dispatch
2020-06-02T13:20:34.445 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service holdoff time over, scheduling restart.
2020-06-02T13:20:34.446 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Stopped Ceph mon.b for 4d893e86-a4d3-11ea-a06b-001a4aab830c.
2020-06-02T13:20:34.447 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Starting Ceph mon.b for 4d893e86-a4d3-11ea-a06b-001a4aab830c...
2020-06-02T13:20:34.586 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 podman[10871]: Error: no container with name or ID ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c-mon.b found: no such container
2020-06-02T13:20:34.595 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Started Ceph mon.b for 4d893e86-a4d3-11ea-a06b-001a4aab830c.
2020-06-02T13:20:34.839 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 bash[10891]: Error: error creating container storage: the container name "ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c-mon.b" is already in use by "15d8802093
344a7c2982ce53c9b435eebf7267d38b890a78522bd74226335309". You have to remove that container to be able to reuse that name.: that name is already in use
2020-06-02T13:20:34.845 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service: main process exited, code=exited, status=125/n/a
2020-06-02T13:20:34.952 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 podman[10911]: Error: no container with name or ID ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c-mon.b found: no such container
2020-06-02T13:20:34.963 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Unit ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service entered failed state.
2020-06-02T13:20:34.964 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service failed.
2020-06-02T13:20:35.017 INFO:tasks.cephadm:Waiting for 2 mons in monmap...
```

/a/kchai-2020-06-02_10:41:24-rados-wip-kefu-testing-2020-06-02-1412-distro-basic-smithi/5111773

not sure if it's related.

**Comment by sebastian-philipp** [comment]
> ```
> 2020-06-02T13:20:34.213 INFO:ceph.mon.a.smithi013.stdout:Jun 02 13:20:34 smithi013 bash[7471]: audit 2020-06-02T13:20:33.431015+0000 mon.a (mon.0) 177 : audit [DBG] from='client.? v1:172.21.15.79:0/3533396549' entity='client.admin' cmd=[{
> "prefix": "mon dump", "format": "json"}]: dispatch
> 2020-06-02T13:20:34.445 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service holdoff time over, scheduling restart.
> 2020-06-02T13:20:34.446 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Stopped Ceph mon.b for 4d893e86-a4d3-11ea-a06b-001a4aab830c.
> 2020-06-02T13:20:34.447 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Starting Ceph mon.b for 4d893e86-a4d3-11ea-a06b-001a4aab830c...
> 2020-06-02T13:20:34.586 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 podman[10871]: Error: no container with name or ID ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c-mon.b found: no such container
> 2020-06-02T13:20:34.595 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Started Ceph mon.b for 4d893e86-a4d3-11ea-a06b-001a4aab830c.
> 2020-06-02T13:20:34.839 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 bash[10891]: Error: error creating container storage: the container name "ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c-mon.b" is already in use by "15d8802093
> 344a7c2982ce53c9b435eebf7267d38b890a78522bd74226335309". You have to remove that container to be able to reuse that name.: that name is already in use
> 2020-06-02T13:20:34.845 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service: main process exited, code=exited, status=125/n/a
> 2020-06-02T13:20:34.952 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 podman[10911]: Error: no container with name or ID ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c-mon.b found: no such container
> 2020-06-02T13:20:34.963 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: Unit ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service entered failed state.
> 2020-06-02T13:20:34.964 INFO:ceph.mon.b.smithi079.stdout:Jun 02 13:20:34 smithi079 systemd[1]: ceph-4d893e86-a4d3-11ea-a06b-001a4aab830c@mon.b.service failed.
> 2020-06-02T13:20:35.017 INFO:tasks.cephadm:Waiting for 2 mons in monmap...
> ```
> 
> /a/kchai-2020-06-02_10:41:24-rados-wip-kefu-testing-2020-06-02-1412-distro-basic-smithi/5111773
> 
> not sure if it's related.

```
Error: container_linux.go:345: starting container process caused "exec: \"/usr/bin/ceph-mon\": stat /usr/bin/ceph-mon: no such file or directory"
```

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:1317]: ping @jmolmo

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:1317]: ping @jmolmo

**Comment by tchaikov** [comment]
> Error: container_linux.go:345: starting container process caused "exec: \"/usr/bin/ceph-mon\": stat /usr/bin/ceph-mon: no such file or directory"


sorry , @sebastian-philipp i am not sure if i understand this. do you mean this is not related? if it's not related. i'd say this is a blocker. as we cannot use cephadm test suite to verify the PRs anymore.

**Comment by sebastian-philipp** [comment]
yep, that's tracked in https://tracker.ceph.com/issues/44990

**Comment by tchaikov** [comment]
@sebastian-philipp thanks for pointing me to that PR. agreed. it's a pain. i commented at https://github.com/ceph/ceph/pull/35018#discussion_r435152379. HTH.

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:1317]: ping @jmolmo

**Pull request merged**